Publish on pypi

This is a quick guide to get it published. you can test everything on testpypi if you do not want to mess things up too 

* create an account on pypi --> pypi/register
* tag master and release it as v0.1 --> releases
* checkout tag v0.1 git checkout v0.1
* do some testing (code)
* (optional) publish to testpypi for testing
* publish to productive pypi (make sure setup.py version is up2date with github release version ;))
  * (first time) python3 setup.py register and save your pypi profile to ~/.pypirc
  * (for every new release version) python3 setup.py sdist upload -r pypi
* test python -m pip install evmlab and python -m evmlab

Courtesy of @tintinweb

Doing the first upload command results in 
```
Registering Evmlab to https://upload.pypi.org/legacy/
Server response (410): Project pre-registration is no longer required or supported, upload your files instead.
```
The second results in 
```
[user@work evmlab]$ python3 setup.py sdist upload -r pypi
/usr/lib64/python3.6/distutils/dist.py:261: UserWarning: Unknown distribution option: 'install_requires'
  warnings.warn(msg)
/usr/lib64/python3.6/distutils/dist.py:261: UserWarning: Unknown distribution option: 'extras_require'
  warnings.warn(msg)
running sdist
running check
warning: sdist: manifest template 'MANIFEST.in' does not exist (using default file list)

warning: sdist: standard file not found: should have one of README, README.txt

writing manifest file 'MANIFEST'
creating Evmlab-0.3
creating Evmlab-0.3/evmlab
creating Evmlab-0.3/evmlab/tools
creating Evmlab-0.3/evmlab/tools/reproducer
creating Evmlab-0.3/evmlab/tools/reproducer/templates
making hard links in Evmlab-0.3...
hard linking setup.py -> Evmlab-0.3
hard linking evmlab/__init__.py -> Evmlab-0.3/evmlab
hard linking evmlab/__main__.py -> Evmlab-0.3/evmlab
hard linking evmlab/compiler.py -> Evmlab-0.3/evmlab
hard linking evmlab/context.py -> Evmlab-0.3/evmlab
hard linking evmlab/contract.py -> Evmlab-0.3/evmlab
hard linking evmlab/etherchain.py -> Evmlab-0.3/evmlab
hard linking evmlab/evmtrace.py -> Evmlab-0.3/evmlab
hard linking evmlab/genesis.py -> Evmlab-0.3/evmlab
hard linking evmlab/multiapi.py -> Evmlab-0.3/evmlab
hard linking evmlab/opcodes.py -> Evmlab-0.3/evmlab
hard linking evmlab/reproduce.py -> Evmlab-0.3/evmlab
hard linking evmlab/source_map.py -> Evmlab-0.3/evmlab
hard linking evmlab/utils.py -> Evmlab-0.3/evmlab
hard linking evmlab/vm.py -> Evmlab-0.3/evmlab
hard linking evmlab/tools/__init__.py -> Evmlab-0.3/evmlab/tools
hard linking evmlab/tools/opviewer.py -> Evmlab-0.3/evmlab/tools
hard linking evmlab/tools/reproducer/__init__.py -> Evmlab-0.3/evmlab/tools/reproducer
hard linking evmlab/tools/reproducer/reproducer.py -> Evmlab-0.3/evmlab/tools/reproducer
hard linking evmlab/tools/reproducer/templates/index.html -> Evmlab-0.3/evmlab/tools/reproducer/templates
creating dist
Creating tar archive
removing 'Evmlab-0.3' (and everything under it)
running upload
Traceback (most recent call last):
  File "setup.py", line 23, in <module>
    "docker": ["docker==3.0.0"]}
  File "/usr/lib64/python3.6/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib64/python3.6/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/usr/lib64/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/usr/lib64/python3.6/distutils/command/upload.py", line 64, in run
    self.upload_file(command, pyversion, filename)
  File "/usr/lib64/python3.6/distutils/command/upload.py", line 74, in upload_file
    raise AssertionError("unsupported schema " + schema)
AssertionError: unsupported schema 
```

--- a/evmlab/tools/opviewer.py
+++ b/evmlab/tools/opviewer.py
@@ -776,7 +776,7 @@ class EvmTrace(object):
     Main Wrapper class to handle Evm Traces
     """
 
-    def __init__(self, api="https://mainnet.infura.io/evmlab"):
+    def __init__(self, api="https://mainnet.infura.io/remix"):
 
         self.api = utils.getApi(api)
         self.source_path = None
@@ -1023,8 +1023,8 @@ python3 opviewer.py -f example.json -s /path/to/contracts -j /path/to/combined.j
 
     web3settings = parser.add_argument_group('Web3',
                                              'Settings about where to fetch information from when displaying contract sources (default infura)')
-    web3settings.add_argument("--web3", type=str, default="https://mainnet.infura.io/",
-                              help="Web3 API url to fetch info from (default 'https://mainnet.infura.io/'")
+    web3settings.add_argument("--web3", type=str, default="https://mainnet.infura.io/remix",
+                              help="Web3 API url to fetch info from (default 'https://mainnet.infura.io/remix'")
 
     args = parser.parse_args()
 

--- a/setup.py
+++ b/setup.py
@@ -1,8 +1,24 @@
 #!/usr/bin/env python
+"""
+deps:
+*) pip install --upgrade pip wheel setuptools twine
+publish:
+1) #> python setup.py sdist bdist_wheel
+2) #> twine upload dist/*   #<specific specific wheel if needed>; --repository <testpypi> or  --repository-url <testpypi-url>
+"""
+import os
+from setuptools import setup
 
-from distutils.core import setup
 
-version = '0.1'
+def read(fname):
+    fpath = os.path.join(os.path.dirname(__file__), fname)
+    fd = open(fpath, 'r')
+    data = fd.read()
+    fd.close()
+    return data
+
+
+version = '0.3.1'  # keep this in sync with the last release tag!
 
 setup(name='Evmlab',
       version=version,
@@ -12,11 +28,15 @@ setup(name='Evmlab',
       license="GPLv3",
       keywords=["ethereum", "transaction", "debugger"],
       url="https://github.com/ethereum/evmlab/",
-      download_url="https://github.com/ethereum/evmlab/tarball/v%s"%version,
+      download_url="https://github.com/ethereum/evmlab/tarball/v%s" % version,
+      # generate rst from .md:  pandoc --from=markdown --to=rst README.md -o README.rst (fix diff section and footer)
+      long_description=read("README.md") if os.path.isfile("README.md") else "",
+      # for pypi.org; empty because we do not ship Readme.md with the package. may otherwise fail on install
+      long_description_content_type='text/markdown',  # requires twine and recent setuptools
       packages=['evmlab',
                 'evmlab.tools',
                 'evmlab.tools.reproducer'],
-      package_data={'evmlab.tools.reproducer':["templates/*"]},
+      package_data={'evmlab.tools.reproducer': ["templates/*"]},
       install_requires=["requests", "web3", "eth-hash[pycryptodome]", "rlp>=1.0"],
       extras_require={"consolegui": ["urwid"],
                       "abidecoder": ["ethereum-input-decoder"],

